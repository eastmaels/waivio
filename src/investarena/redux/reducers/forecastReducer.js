import moment from 'moment';
import * as activeForecastTypes from '../actions/forecastActions';

const initialState = {
  forecastData: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case activeForecastTypes.GET_FORECAST_DATA.SUCCESS:
      return {
        ...state,
        forecastData: action.payload.forecasts.sort(
          (a, b) => moment(b.created_at).unix() - moment(a.created_at).unix(),
        ), // eslint-disable-line
      };
    case activeForecastTypes.GET_FORECAST_DATA.ERROR:
      return {
        ...state,
        forecastData: [],
      };

    default:
      return state;
  }
};

export const getForecastData = state => state.forecastData;
