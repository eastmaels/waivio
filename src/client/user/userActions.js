import { createAction } from 'redux-actions';
import { getIsAuthenticated, getAuthenticatedUserName } from '../reducers';
import { getAllFollowing } from '../helpers/apiHelpers';
import { createAsyncActionType } from '../helpers/stateHelpers';
import * as ApiClient from '../../waivioApi/ApiClient';
import { getUserCoordinatesByIpAdress } from '../components/Maps/mapHelper';

require('isomorphic-fetch');

export const FOLLOW_USER = '@user/FOLLOW_USER';
export const FOLLOW_USER_START = '@user/FOLLOW_USER_START';
export const FOLLOW_USER_SUCCESS = '@user/FOLLOW_USER_SUCCESS';
export const FOLLOW_USER_ERROR = '@user/FOLLOW_USER_ERROR';

export const followUser = username => (dispatch, getState, { steemConnectAPI }) => {
  const state = getState();

  if (!getIsAuthenticated(state)) {
    return Promise.reject('User is not authenticated');
  }

  return dispatch({
    type: FOLLOW_USER,
    payload: {
      promise: steemConnectAPI.follow(getAuthenticatedUserName(state), username),
    },
    meta: {
      username,
    },
  });
};

export const GET_RECOMMENDED_OBJECTS = '@user/GET_RECOMMENDED_OBJECTS';
export const GET_RECOMMENDED_OBJECTS_START = '@user/GET_RECOMMENDED_OBJECTS_START';
export const GET_RECOMMENDED_OBJECTS_SUCCESS = '@user/GET_RECOMMENDED_OBJECTS_SUCCESS';
export const GET_RECOMMENDED_OBJECTS_ERROR = '@user/GET_RECOMMENDED_OBJECTS_ERROR';

export const getRecommendedObj = () => dispatch => {
  return dispatch({
    type: GET_RECOMMENDED_OBJECTS,
    payload: {
      promise: ApiClient.getRecommendedObjects(),
    },
  });
};

export const UNFOLLOW_USER = '@user/UNFOLLOW_USER';
export const UNFOLLOW_USER_START = '@user/UNFOLLOW_USER_START';
export const UNFOLLOW_USER_SUCCESS = '@user/UNFOLLOW_USER_SUCCESS';
export const UNFOLLOW_USER_ERROR = '@user/UNFOLLOW_USER_ERROR';

export const unfollowUser = username => (dispatch, getState, { steemConnectAPI }) => {
  const state = getState();

  if (!getIsAuthenticated(state)) {
    return Promise.reject('User is not authenticated');
  }

  return dispatch({
    type: UNFOLLOW_USER,
    payload: {
      promise: steemConnectAPI.unfollow(getAuthenticatedUserName(state), username),
    },
    meta: {
      username,
    },
  });
};

export const GET_FOLLOWING = '@user/GET_FOLLOWING';
export const GET_FOLLOWING_START = '@user/GET_FOLLOWING_START';
export const GET_FOLLOWING_SUCCESS = '@user/GET_FOLLOWING_SUCCESS';
export const GET_FOLLOWING_ERROR = '@user/GET_FOLLOWING_ERROR';

export const getFollowing = username => (dispatch, getState) => {
  const state = getState();

  if (!username && !getIsAuthenticated(state)) {
    return dispatch({ type: GET_FOLLOWING_ERROR });
  }

  const targetUsername = username || getAuthenticatedUserName(state);

  return dispatch({
    type: GET_FOLLOWING,
    meta: targetUsername,
    payload: {
      promise: getAllFollowing(targetUsername),
    },
  });
};

export const GET_FOLLOWING_OBJECTS = '@user/GET_FOLLOWING_OBJECTS';
export const GET_FOLLOWING_OBJECTS_START = '@user/GET_FOLLOWING_OBJECTS_START';
export const GET_FOLLOWING_OBJECTS_SUCCESS = '@user/GET_FOLLOWING_OBJECTS_SUCCESS';
export const GET_FOLLOWING_OBJECTS_ERROR = '@user/GET_FOLLOWING_OBJECTS_ERROR';

export const getFollowingObjects = username => (dispatch, getState) => {
  const state = getState();

  if (!username && !getIsAuthenticated(state)) {
    return dispatch({ type: GET_FOLLOWING_ERROR });
  }

  const targetUsername = username || getAuthenticatedUserName(state);
  return dispatch({
    type: GET_FOLLOWING_OBJECTS,
    payload: {
      promise: ApiClient.getAllFollowingObjects(targetUsername),
    },
  });
};

export const UPDATE_RECOMMENDATIONS = '@user/UPDATE_RECOMMENDATIONS';
export const updateRecommendations = createAction(UPDATE_RECOMMENDATIONS);

export const GET_NOTIFICATIONS = createAsyncActionType('@user/GET_NOTIFICATIONS');

export const getNotifications = username => (dispatch, getState, { busyAPI }) => {
  const state = getState();

  if (!username && !getIsAuthenticated(state)) {
    return dispatch({ type: GET_NOTIFICATIONS.ERROR });
  }

  const targetUsername = username || getAuthenticatedUserName(state);

  return dispatch({
    type: GET_NOTIFICATIONS.ACTION,
    meta: targetUsername,
    payload: {
      promise: busyAPI.sendAsync('get_notifications', [targetUsername]),
    },
  });
};

export const GET_USER_LOCATION = createAsyncActionType('@user/GET_USER_LOCATION');

export const getCoordinates = () => dispatch =>
  dispatch({
    type: GET_USER_LOCATION.ACTION,
    payload: getUserCoordinatesByIpAdress(),
  });
