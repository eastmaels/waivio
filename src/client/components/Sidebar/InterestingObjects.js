import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import ObjectCard from './ObjectCard';
import './InterestingObjects.less';
import './SidebarContentBlock.less';
import { getRecommendedObjects } from '../../reducers';
import { getRecommendedObj } from '../../user/userActions';
import RightSidebarLoading from '../../app/Sidebar/RightSidebarLoading';

@connect(
  state => ({
    recommendedObjects: getRecommendedObjects(state),
  }),
  {
    getRecommendedObj,
  },
)
class InterestingObjects extends React.Component {
  static propTypes = {
    recommendedObjects: PropTypes.arrayOf(PropTypes.shape({ tag: PropTypes.string })).isRequired,
    getRecommendedObj: PropTypes.func.isRequired,
  };
  static defaultProps = {
    objects: [],
  };

  componentWillMount() {
    if (_.size(this.props.recommendedObjects) < 5) {
      this.props.getRecommendedObj();
    }
  }

  render() {
    const { recommendedObjects } = this.props;
    return _.size(recommendedObjects) >= 5 ? (
      <div className="InterestingObjects SidebarContentBlock">
        <h4 className="SidebarContentBlock__title">
          <i className="iconfont icon-collection SidebarContentBlock__icon" />{' '}
          <FormattedMessage id="interesting_objects" defaultMessage="Top 5 Objects" />
        </h4>
        <div className="SidebarContentBlock__content">
          {recommendedObjects &&
            recommendedObjects.map(wobject => (
              <ObjectCard key={wobject.author_permlink} wobject={wobject} />
            ))}
          <h4 className="InterestingObjects__more">
            <Link to={'/objects'}>
              <FormattedMessage id="discover_more_objects" defaultMessage="Discover more objects" />
            </Link>
          </h4>
        </div>
      </div>
    ) : (
      <RightSidebarLoading />
    );
  }
}
export default InterestingObjects;
