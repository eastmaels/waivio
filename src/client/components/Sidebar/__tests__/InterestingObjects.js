import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import React from 'react';
import InterestingObjects from '../InterestingObjects';
import * as reducers from '../../../reducers';
import * as actions from '../../../user/userActions';
import { shallowWithStore, mountWithStore } from '../__tests__/shallowWrapper';

const mockStore = configureMockStore([thunk]);
const store = mockStore({});

jest.mock('../../../../waivioApi/ApiClient');

describe('<InterestingObjects />', () => {
  const propsComponent = {
    recommendedObjects: [
      {
        __v: 0,
        _id: '5caf424c6fb6c810cde66d6b',
        app: 'waiviodev/1.0.0',
        author: 'suy38',
        author_permlink: 'life',
        community: '',
        createdAt: '2019-04-11T13:34:04.222Z',
        creator: 'wiv01',
        default_name: 'life',
        is_extending_open: true,
        is_posting_open: true,
        object_type: 'hashtag',
        parent: '',
        rank: 8,
        updatedAt: '2019-06-18T11:33:16.621Z',
        user_count: 5,
        weight: 32967838660.60074,
      },
      {
        __v: 0,
        _id: '5caf43596fb6c810cde6b716',
        app: 'waiviodev/1.0.0',
        author: 'q1w2c',
        author_permlink: 'busy',
        community: '',
        createdAt: '2019-04-11T13:38:33.125Z',
        creator: 'wiv01',
        default_name: 'busy',
        is_extending_open: true,
        is_posting_open: true,
        object_type: 'hashtag',
        parent: '',
        rank: 7,
        updatedAt: '2019-06-18T11:33:17.199Z',
        user_count: 5,
        weight: 30582410090.73522,
      },
    ],
  };

  it('renders component without crashing', () => {
    const component = shallowWithStore(<InterestingObjects {...propsComponent} />, store);
    expect(component).toMatchSnapshot();
  });

  it('should render page if quote exist', () => {
    reducers.getRecommendedObjects = jest.fn(() => new Array(5));
    const component = mountWithStore(<InterestingObjects {...propsComponent} />, store);
    const mainDiv = component.find('.InterestingObjects');
    expect(mainDiv).toHaveLength(1);
  });

  it('should not call getRecommendedObj if recomended objects not less then 5', async () => {
    reducers.getRecommendedObjects = jest.fn(() => new Array(5));
    actions.getRecommendedObj = jest.fn(() => Promise.resolve());
    const mounted = mountWithStore(<InterestingObjects />, store);
    expect(actions.getRecommendedObj).toHaveBeenCalledTimes(0);
  });

  it('should call getRecommendedObj if recomended objects less then 5', async () => {
    const mockGetRecommendedObj = jest.fn(() => Promise.resolve());
    reducers.getRecommendedObjects = jest.fn(() => new Array(4));
    mountWithStore(<InterestingObjects />, store);
    expect(mockGetRecommendedObj).toHaveBeenCalledTimes(1);
  });
});
