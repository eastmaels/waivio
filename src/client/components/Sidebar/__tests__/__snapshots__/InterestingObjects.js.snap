// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InterestingObjects /> renders component without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <IntlProvider
      locale="en"
    >
      <Connect(InterestingObjects)
        recommendedObjects={
          Array [
            Object {
              "__v": 0,
              "_id": "5caf424c6fb6c810cde66d6b",
              "app": "waiviodev/1.0.0",
              "author": "suy38",
              "author_permlink": "life",
              "community": "",
              "createdAt": "2019-04-11T13:34:04.222Z",
              "creator": "wiv01",
              "default_name": "life",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 8,
              "updatedAt": "2019-06-18T11:33:16.621Z",
              "user_count": 5,
              "weight": 32967838660.60074,
            },
            Object {
              "__v": 0,
              "_id": "5caf43596fb6c810cde6b716",
              "app": "waiviodev/1.0.0",
              "author": "q1w2c",
              "author_permlink": "busy",
              "community": "",
              "createdAt": "2019-04-11T13:38:33.125Z",
              "creator": "wiv01",
              "default_name": "busy",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 7,
              "updatedAt": "2019-06-18T11:33:17.199Z",
              "user_count": 5,
              "weight": 30582410090.73522,
            },
          ]
        }
      />
    </IntlProvider>
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Connect(InterestingObjects)
        recommendedObjects={
          Array [
            Object {
              "__v": 0,
              "_id": "5caf424c6fb6c810cde66d6b",
              "app": "waiviodev/1.0.0",
              "author": "suy38",
              "author_permlink": "life",
              "community": "",
              "createdAt": "2019-04-11T13:34:04.222Z",
              "creator": "wiv01",
              "default_name": "life",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 8,
              "updatedAt": "2019-06-18T11:33:16.621Z",
              "user_count": 5,
              "weight": 32967838660.60074,
            },
            Object {
              "__v": 0,
              "_id": "5caf43596fb6c810cde6b716",
              "app": "waiviodev/1.0.0",
              "author": "q1w2c",
              "author_permlink": "busy",
              "community": "",
              "createdAt": "2019-04-11T13:38:33.125Z",
              "creator": "wiv01",
              "default_name": "busy",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 7,
              "updatedAt": "2019-06-18T11:33:17.199Z",
              "user_count": 5,
              "weight": 30582410090.73522,
            },
          ]
        }
      />,
      "locale": "en",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "recommendedObjects": Array [
          Object {
            "__v": 0,
            "_id": "5caf424c6fb6c810cde66d6b",
            "app": "waiviodev/1.0.0",
            "author": "suy38",
            "author_permlink": "life",
            "community": "",
            "createdAt": "2019-04-11T13:34:04.222Z",
            "creator": "wiv01",
            "default_name": "life",
            "is_extending_open": true,
            "is_posting_open": true,
            "object_type": "hashtag",
            "parent": "",
            "rank": 8,
            "updatedAt": "2019-06-18T11:33:16.621Z",
            "user_count": 5,
            "weight": 32967838660.60074,
          },
          Object {
            "__v": 0,
            "_id": "5caf43596fb6c810cde6b716",
            "app": "waiviodev/1.0.0",
            "author": "q1w2c",
            "author_permlink": "busy",
            "community": "",
            "createdAt": "2019-04-11T13:38:33.125Z",
            "creator": "wiv01",
            "default_name": "busy",
            "is_extending_open": true,
            "is_posting_open": true,
            "object_type": "hashtag",
            "parent": "",
            "rank": 7,
            "updatedAt": "2019-06-18T11:33:17.199Z",
            "user_count": 5,
            "weight": 30582410090.73522,
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Connect(InterestingObjects)
          recommendedObjects={
            Array [
              Object {
                "__v": 0,
                "_id": "5caf424c6fb6c810cde66d6b",
                "app": "waiviodev/1.0.0",
                "author": "suy38",
                "author_permlink": "life",
                "community": "",
                "createdAt": "2019-04-11T13:34:04.222Z",
                "creator": "wiv01",
                "default_name": "life",
                "is_extending_open": true,
                "is_posting_open": true,
                "object_type": "hashtag",
                "parent": "",
                "rank": 8,
                "updatedAt": "2019-06-18T11:33:16.621Z",
                "user_count": 5,
                "weight": 32967838660.60074,
              },
              Object {
                "__v": 0,
                "_id": "5caf43596fb6c810cde6b716",
                "app": "waiviodev/1.0.0",
                "author": "q1w2c",
                "author_permlink": "busy",
                "community": "",
                "createdAt": "2019-04-11T13:38:33.125Z",
                "creator": "wiv01",
                "default_name": "busy",
                "is_extending_open": true,
                "is_posting_open": true,
                "object_type": "hashtag",
                "parent": "",
                "rank": 7,
                "updatedAt": "2019-06-18T11:33:17.199Z",
                "user_count": 5,
                "weight": 30582410090.73522,
              },
            ]
          }
        />,
        "locale": "en",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "recommendedObjects": Array [
            Object {
              "__v": 0,
              "_id": "5caf424c6fb6c810cde66d6b",
              "app": "waiviodev/1.0.0",
              "author": "suy38",
              "author_permlink": "life",
              "community": "",
              "createdAt": "2019-04-11T13:34:04.222Z",
              "creator": "wiv01",
              "default_name": "life",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 8,
              "updatedAt": "2019-06-18T11:33:16.621Z",
              "user_count": 5,
              "weight": 32967838660.60074,
            },
            Object {
              "__v": 0,
              "_id": "5caf43596fb6c810cde6b716",
              "app": "waiviodev/1.0.0",
              "author": "q1w2c",
              "author_permlink": "busy",
              "community": "",
              "createdAt": "2019-04-11T13:38:33.125Z",
              "creator": "wiv01",
              "default_name": "busy",
              "is_extending_open": true,
              "is_posting_open": true,
              "object_type": "hashtag",
              "parent": "",
              "rank": 7,
              "updatedAt": "2019-06-18T11:33:17.199Z",
              "user_count": 5,
              "weight": 30582410090.73522,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
